### PR TITLE
feat: ping-pong services

### PR DESCRIPTION
**Pull request by perenesenko**
Fix for: https://github.com/argoproj/argo-rollouts/issues/1453

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: docs/features/kustomize/rollout_cr_schema.json [hunk 1]
```diff
@@ -47,6 +47,22 @@
                                                     "type": "array",
                                                     "x-kubernetes-patch-merge-key": "name",
                                                     "x-kubernetes-patch-strategy": "merge"
+                                                },
```
Comment (harikrongali) [inline docs/features/kustomize/rollout_cr_schema.json:50]: why is dry run showing in teh diff?

File: rollout/trafficrouting/service_helper.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			return canary.PingPong.PingService, canary.PingPong.PongService
+		} else {
+			return canary.PingPong.PongService, canary.PingPong.PingService
+		}
+	} else {
+		return ro.Spec.Strategy.Canary.StableService, ro.Spec.Strategy.Canary.CanaryService
+	}
+}
+
+func IsStablePing(ro *v1alpha1.Rollout) bool { // reurn "stable is ping"
```
Comment (harikrongali) [inline rollout/trafficrouting/service_helper.go:23]: remove the comment

File: rollout/trafficrouting/service_helper.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	return ro.Status.Canary.StablePingPong == v1alpha1.PPPing
+}
+
+func PingPongOpposite(ro *v1alpha1.Rollout) v1alpha1.PingPongType {
+	//status.stablePingPong == ""
```
Comment (harikrongali) [inline rollout/trafficrouting/service_helper.go:28]: same here...
can you add a comment block giving different scenarios and  expected return value

File: docs/features/kustomize/rollout_cr_schema.json (hunk 1 above)
Comment (perenesenko) [reply docs/features/kustomize/rollout_cr_schema.json:50]: I guess previously the `codegen` for dryRun was skipped

File: rollout/trafficrouting/service_helper.go (hunk 2 above)
Comment (perenesenko) [reply rollout/trafficrouting/service_helper.go]: done

File: rollout/trafficrouting/service_helper.go (hunk 3 above)
Comment (perenesenko) [reply rollout/trafficrouting/service_helper.go]: This method is used in one place only so I've inlined it.

**Review by harikrongali — APPROVED** [review]

**Review by jessesuen — COMMENTED** [review]
Minor request to ensure this is not used with other traffic routers.

File: pkg/apis/rollouts/validation/validation.go [hunk 4]
```diff
@@ -224,6 +230,20 @@ func ValidateRolloutStrategyCanary(rollout *v1alpha1.Rollout, fldPath *field.Pat
...
+		}
+		if canary.PingPong.PongService == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("pingPong").Child("pongService"), canary.PingPong.PongService, InvalidPingPongProvidedMessage))
+		}
+		if canary.PingPong.PingService == canary.PingPong.PongService {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("pingPong").Child("pingService"), canary.PingPong.PingService, DuplicatedPingPongServicesMessage))
+		}
+		if canary.TrafficRouting != nil && canary.TrafficRouting.ALB != nil && canary.TrafficRouting.ALB.RootService == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("trafficRouting").Child("alb").Child("rootService"), canary.TrafficRouting.ALB.RootService, MissedAlbRootServiceMessage))
+		}
```
[the remark this answers is not in the document]
Comment (perenesenko) [reply pkg/apis/rollouts/validation/validation.go:250]: Done

**Review by jessesuen — APPROVED** [review]
Great job!

**Pull request merged**